### ISSUE TITLE
Empty block/region detection 

### ISSUE THREAD

**Issue by erikerskine** [localgovdrupal/localgov_base#597]
## TL;DR:

We cannot have both these things:
- guaranteed removal of empty regions
- lazy rendering of blocks (including BigPipe, placeholders for caching etc)

The current implementation tries to do both and is broken.

Pick one of those things. Perhaps on a per-block or per-region basis.

## Summary

There's a long and tricky [core issue](https://www.drupal.org/node/953034) about this. Here's a distilled version & how it applies to `localgov_base`:

1. Drupal renders page templates before individual blocks. That's good - it facilitates BigPipe, and means that mostly stable pages with small amounts of volatile content can be cached effectively.

2. At the time the page is rendered we know what blocks are present, but we don't know what they will contain (if anything). Here is the conundrum: we want to know if a block contains _something_, because we want to use classes like `has-sidebar` in `page.html.twig`. But all we have is a placeholder, not actual content.

3. If we do nothing we must accept that a block (and a region) might end being empty. **In this case "has sidebar" really means "might have a sidebar"**.

4. Anything more complicated involves choosing between guaranteed detection of "is empty" and lazy building.

### Choosing lazy building

This is the default, so we don't need to do anything.

We can mitigate against empty blocks a bit with condition plugins. Conditions are evaluated ahead of time. For example: a view that only produces results for logged-in users can be hidden from anonymous users when it is placed. There is no need to render it and examine the output.

### Choosing guaranteed "is empty" detection

This involves sacrificing lazy building.

There are various approaches suggested online. All some variation of the following, done in a `hook_preprocess_page()`:
1. find the placeholder for a block
2. render that block
3. replace the placeholder with what we just rendered
4. apply the cache metadata to the page

Points 2/3 render BigPipe useless.

Point 4 reduces cacheability. It's particularly agressive with hard-to-cache or uncacheable blocks: they make the whole page equally hard to cache.

The current solution in `localgov_base` does steps 1-3 only.

## Demonstrating the issue

Here is a block that displays a message when the current time is an even number of minutes.

It's a really contrived example to make it obvious what's going on. But this could just as easily be an uncacheable view, or some other dynamic thing.

Place the block in a sidebar.

```php
final class FunnyBlock extends BlockBase {

  // This block can never be cached.
  use UncacheableDependencyTrait;

  public function build() {
    // We only show content during even numbered minutes.
    $seconds = strtotime('now');
    $minutes = intdiv($seconds, 60);
    $is_even_numbered_minute = ($minutes % 2 === 0);

    $build = [];

    if ($is_even_numbered_minute) {
      $build['message'] = ['#plain_text' => 'Even numbered minute.'];
    }

    return $build;
  }

}
```

### Expected behaviour with lazy building

1. During even numbered minutes, there is a sidebar and text
2. During odd numbered minutes, there is a sidebar but no text
3. The page template is cachable

The sidebar is shown because we see the placeholder inside the sidebar and deem that `has-sidebar` is appropriate.

### Expected behaviour with guaranteed "is empty" detection

1. During even numbered minutes, there is a sidebar and text
2. During odd numbered minutes, there is neither sidebar nor text
3. The page template is not cachable.

The page is not cacheable because `has-sidebar` is not cacheable.

### Actual behaviour

localgov_base tries to guarantee empty/non-empty detection. But here's what happens:

1. Start with an otherwise cacheable page. A node will do.
2. Place the block in the sidebar.
3. Clear cache.
4. Visit the page during an even numbered minute: the sidebar and block appear (good).
5. Visit the page again one minute later: there is an empty sidebar (bad).
6. Clear cache.
7. Visit the page during an odd numbered minute: no sidebar, no block (good).
8. Visit the page again one minute later: still no sidebar or block (bad).

Why this fails: in step 4 and 7 we did not apply the cacheability of the rendered block to the page. So the page was cached with whatever `has-sidebar` value it had at the time. No consideration was made that `has-sidebar` might change.

## Steps forward

Before we can do anything, can we establish how should this theme actually behave?

- Guarantee emptiness detection for all blocks/regions, at the expense of caching
- Don't guarantee it, accept that "has sidebar" means "might have sidebar"
- Do it sometimes? If so, which blocks/regions should get this treatment? Make it opt-in?
- Something else?

**Comment by markconroy** [comment]
Fantastic issue @erikerskine.

Thanks for taking the time to document it.

I think this might make a good topic at the Tech Group Drop-in on Thursday if you wanted to come along to it.

I'd like to hear other people's thoughts, but I have a feeling 'Don't guarantee it, accept that "has sidebar" means "might have sidebar"' might need to be the winner, and some `:has` CSS to show/hide the sidebar when in it's "might have" state.

@finnlewis @stephen-cox @ekes I'd like your thoughts on this if you have some time to think about it.

---
Thanks to [Big Blue Door](https://www.bigbluedoor.net/) for sponsoring my time to work on this.

**Comment by andybroomfield** [comment]
The main thing is actually detecting the sidebar. I don't know about localgov_base but the BHCC base theme does need to know if a sidebar is present to set certian layout classes. Maybe there is a way to detect if the sidebar is then genuinely empty in css.

FWIW Brighton has been disabling this logic on other regions except the sidebar so we can benefit from lazy loading. The system messages being displayed being the most notable.

**Comment by ctorgalson** [comment]
@andybroomfield, @markconroy you anticipated my post :)

A little additional info re: the "choosing lazy building" option.

When @erikerskine first pointed this out to us, I did a test to address the problem with CSS only. Given the existence of [`:empty`](https://developer.mozilla.org/en-US/docs/Web/CSS/:empty), [`:blank`](https://developer.mozilla.org/en-US/docs/Web/CSS/:blank), and [`:has()`](https://developer.mozilla.org/en-US/docs/Web/CSS/:has), I thought the tools to do something should be available.

What I found:

-  `:empty` won't do as it only checks that there are no _element_ children; text nodes like empty space don't match `:empty`
- `:blank` won't do as it's not yet well-supported
- `:has()` can be made to work with no real problems beyond a slight template rewrite

I'd have to spend time recreating my test in LGD-specific css/Twig, but I put together [a proof-of-concept codepen to show the basic concept](https://codepen.io/bedlamhotel/pen/JjQZxqz?editors=0110), but the TL;DR: is that it's not very difficult.

The markup for both sidebars is present in *all* of the boxes in the screenshot. Note that the coloured outlines indicate that the same selectors that hide the sidebar can target the overall layout as well.

![image](https://github.com/user-attachments/assets/d6d93d56-b3df-4af1-be21-767ccd5a3120)

**Comment by andybroomfield** [comment]
Support for [css has()](https://caniuse.com/css-has) is for browsers from 2022, could that be an issue for what people need to support? Is there gracefull fallback, is rendering the sidebars always ok?

**Comment by ctorgalson** [comment]
> could that be an issue for what people need to support?

In general, I'd think no (Chrome-based browsers and Safari have supported it for some time, and now that Firefox also does, Annertech uses it routinely), but it depends on LGD's overall policy w/r/t to browser support (related question: is that documented somwhere?)

Absent the use of `:has()`, i think we could rule out a passive (i.e. css-only) solution.

---

Added: there _is_ one possibility, but it'd be brittle and could cause confusion in sub-themes even if working in `localgov_base`: if we could _guarantee_ that the sidebars didn't contain whitespace, the use of `:empty` would do. But even one space between the opening and closing tags would cause the sidebar to render...

**Comment by markconroy** [comment]
As far as I remember our browser support policy is the same as Drupal's. And as far as I know, Drupal allows the use of `:has`

**Comment by ctorgalson** [comment]
> As far as I remember our browser support policy is the same as Drupal's. And as far as I know, Drupal allows the use of :has

By the strict "letter of the law", [it looks like it might not :frowning_face:](https://www.drupal.org/docs/getting-started/system-requirements/browser-requirements) since Opera Mini is in the supported list.

![image](https://github.com/user-attachments/assets/0559fa98-217a-42aa-955a-39b1a0633b8a)

**Comment by andybroomfield** [comment]
Noting the [GDS guidelines ](https://www.gov.uk/service-manual/technology/designing-for-different-browsers-and-devices) on what to support. They (as of August 2024) still reccommend Safari 12.

**Comment by erikerskine** [comment]
> I think this might make a good topic at the Tech Group Drop-in on Thursday if you wanted to come along to it.

@markconroy sure, I'll try to join this week.

**Comment by erikerskine** [comment]
CSS-based solutions like `:has` might be a distraction here.

Asynchronous content (via BigPipe) is good for performance/caching, but it means asynchronous DOM changes. I'd be wary of basing layout on that.

Eg this:

`<div class="sidebar"><placeholder-for-some-slow-bigpipe-block/></div>`

can suddenly change to this, long after the main page content is visible:

`<div class="sidebar"></div>`

Maybe it's helpful to answer these two questions first:

- How big a problem is an empty region?
- How many cases can be avoided with better visibility rules in block configuration?

**Comment by ctorgalson** [comment]
> can suddenly change to this, long after the main page content is visible:

It's exactly that case where something like `:has()` would work perfectly--but it looks like css-only options are out due to support requirements in any case.

**Comment by markconroy** [comment]
I don't think CSS-only options are out (yet).

If this fixes a lot, and only affects Opera Mini, I think we could push it through (after a few more chats).

**Comment by ctorgalson** [comment]
@markconroy @andybroomfield pointed out above that Safari 12 is also a blocker.

https://github.com/localgovdrupal/localgov_base/issues/597#issuecomment-2314807283

**Comment by markconroy** [comment]
Are we following GDS or Drupal? Or both? Or our own version of a pick-and-mix and each?

I know we have `:has` in our codebase already, so I don't have an issue with us adding it in again.

Also, in this case, if we have this:

```
<sidebar>
<main>
```

In Safari mobile, that's going to put an empty `<sidebar>` on top of the main rather than beside it, so it will make no difference. On desktop Safari, it might mean space where a sidebar could be on a tiny amount of sites, but the better cachability of things might be worth that trade off.

Let's trash it out more in tech group drop in

**Comment by andybroomfield** [comment]
Its a blocker if you need to follow GDS. I'm trying to find out how current that list is. Its also a question of how we want to set browser support, and if gracefull fallback (presumably a sidebar is always shown) is acceptable.

**Comment by millnut** [comment]
I've been chasing GDS to update that browser support table for quite some time as it should be using Safari 14.5 minimum or even 15.4 (which was released March 15, 2022) by now based on Apple support for their devices. Safari 12.1 was released March 25, 2019 so I would imagine that is well past being supported by Apple.

It still doesn't even list iPadOS as a separate entity from iOS and macOS.

**Comment by erikerskine** [comment]
I analysed the code a bit more to see how we might fix the current logic. There are a couple of things:

```
$copy = $variables['page'][$region];

/** @var \Drupal\Core\Render\RendererInterface $renderer */
$renderer = \Drupal::service('renderer');
$rendered = DeprecationHelper::backwardsCompatibleCall(
  currentVersion: \Drupal::VERSION,
  deprecatedVersion: '10.3',
  currentCallable: fn() => $renderer->renderInIsolation($copy),
  deprecatedCallable: fn() => $renderer->renderPlain($copy),
);
```

1. `$rendered` is being discarded once we've checked for emptiness. But that means we need to render things twice, leading to side-effects (duplicate form submissions, disappearing status messages etc). We can avoid that by replacing the original element with `$rendered`:

```
$variables['page'][$region] = $rendered ? ['#markup' => $rendered] : [];
```

2. After the `renderInIsoloation/renderPlain` call, `$copy` contains cache metadata. This needs to apply to the whole page:

```
$cache_metadata = CacheableMetadata::createFromRenderArray($copy);
$cache_metadata->applyTo($variables);
```

**Comment by erikerskine** [comment]
The proposed fix above is costly.

I'm not sure it should be done everywhere, unconditionally. Nor do I think it's the job of a _base theme_ to decide where.

So here's an idea to opt out of lazy building on a per-block basis. Same logic as above, you just need to enable it for troublesome blocks: https://git.drupalcode.org/sandbox/ingaro-3471436

**Comment by stephen-cox** [comment]
Still trying to get my head around the full implications of this, but it seems clear that calculating the `'has_' . $region` variables is actually working against Drupal's caching and so needs looking at.

It also explains why we've had to exclude some regions from this calculation due to caching issues.

If we can strip out all the has_{region} checks in the templates and replace with a CSS solution then that sounds okay. If not, it sounds like we need to continue to render blocks to check if they're empty. Can we get away with doing this in Twig? It would at least allow subthemes to override this as needed.

**Comment by andybroomfield** [comment]
The two that are important (for layout) are really the two sidebars, as that has implications for the container width. For the others I assume you end up with some empty divs.

**Comment by ctorgalson** [comment]
I looked into handling this in CSS _without_ `:has()`, and it's _nearly_ possible with CSS grids, but not quite (the gap between main content and sidebar remains, even if the sidebar has zero width).

I updated[ the codepen](https://codepen.io/bedlamhotel/pen/JjQZxqz) with a trivially simple way to do this with javascript (which would also be simple to extend with a `ResizeObsever` and/or `MutationObserver` if used). It works well, and is pretty fast.

The downside is it probably is _not_ fast enough. As [a standalone page](https://cdpn.io/bedlamhotel/fullpage/JjQZxqz) it's fast enough that e.g. lighthouse doesn't complain about screen-redrawing, but I doubt that'd be true on a Drupal page full of other js behaviors.

**Comment by millnut** [comment]
I know the GDS has a browser support page but the gov.uk frontend has moved to a grading support scale https://github.com/alphagov/govuk-frontend?tab=readme-ov-file#browser-and-assistive-technology-support and https://github.com/alphagov/govuk-frontend/blob/main/docs/contributing/browser-support.md so I wonder if we could get away with `:has()`

**Comment by erikerskine** [comment]
To test the [sandbox module](https://git.drupalcode.org/sandbox/ingaro-3471436), you need to apply [this MR](https://github.com/localgovdrupal/localgov_base/pull/601)

The MR stops the theme from rendering regions up front. The contrib module does it on an opt-in basis.

This doesn't alter anything that would prevent us also using `:has`.

**Comment by ctorgalson** [comment]
It occurs to me we could just choose an "all of the above" approach to solve the problems of caching and browser support:

- use `has_sidebar` in Twig without doing the render operation (meaning it will _sometimes_ be wrong), _and_
- use `:has()` in css as a fallback (which, for _some_ browsers, will fail), _and_
- use [CSS.supports()](https://developer.mozilla.org/en-US/docs/Web/API/CSS/supports_static) in js to hide empty sidebars if `:has()` isn't supported (which, if javascript is disabled, could fail for _some_ users).

This would mean that in _most_ cases, the layout is handled in Twig (like now), in _most_ of the remaining cases, the layout would be handed by `:has()`, and in _most_ of the very few cases left, the layout is handled by javascript that only runs when needed.

If, in addition, we used some css transition to shrink the sidebar out of the layout, I don't think we'd cause much of a problem with Lighthouse and similar tools.

**Comment by erikerskine** [comment]
@stephen-cox re:

> Still trying to get my head around the full implications of this, but it seems clear that calculating the 'has_' . $region variables is actually working against Drupal's caching and so needs looking at.

Correct.

When performing the calculation, we get some extra cache metadata as well as the empty/not-empty result. Determining layout is more complicated this way, so the total amount of cache metadata needed to do so goes up.

The status quo is to perform the calculation, but discarding the extra metadata from it. The page is incorrectly cached.

One option is to fix the calculation. That means applying the extra metadata to the page. We end up with less cacheable pages.

Another option is to remove the calculation, and loosen the definition of "not empty". This is Drupal's default behaviour. But it is a regression for sites that depend on it.

The sandbox module I proposed would work in conjunction with removal. It puts the calculation back in, but only when necessary. Individual sites/themes need to turn it on when they need to.

> If we can strip out all the has_{region} checks in the templates and replace with a CSS solution then that sounds okay.

Agreed.

> If not [a CSS solution], it sounds like we need to continue to render blocks to check if they're empty.

The question is then how often you do this:
- unconditionally, for every block (safe but costly)
- for all blocks in a particular region (safe but costly, limited in scope)
- only if the block is causing a problem (most efficient, but needs turning on by sitebuilders)

> Can we get away with doing this in Twig? It would at least allow subthemes to override this as needed.

Maybe - we'd need a new twig extension to do `renderInIsolation`. Something like this?

```twig
{% if page.sidebar_first|render_in_isolation is not empty %}
  <div class="sidebar-first">
    {{ page.sidebar_first }}
  </div>
{% endif %}
```

Would you expect subthemes to add that, or remove it? What is the default?

**Comment by markconroy** [comment]
I've spent a decent amount of time working on this this morning trying to figure out what CSS we need for it. 

This seems to work:
```
/* If the sidebar is empty, hide it. */
.lgd-row .sidebar--first:not(:has(.region)),
.lgd-row .sidebar--second:not(:has(.region)) {
  display: none;
}

/* If the sidebar has something in it, show it. */
.lgd-row .sidebar--first:has(.region),
.lgd-row .sidebar--second:has(.region) {
  display: block;
}

/*
  Let the main content grow to fill available space if the sidebars are hidden.
*/
.lgd-row > .sidebar + .lgd-row__one-half {
  flex-grow: 1;
}
```

However, when we get to the `page.html.twig` file and I started removing the clauses for `{% if has_region_name %}` I started to run into issues (I think).

Currently we do a check to see if we have any sidebars. If we do, we place a container around the page so things don't flow to 100% of the screen width, like so pseudo code:
```
{% if has_sidebars %}
  <div class="container">
      <div class="sidebar"></div>
      <div class="main-content"></div>
   </div>
  {% else %}
      <div class="main-content"></div>
{% endif %}
```

We are using the node template to add the container class for nodes, so some nodes or items in nodes can take up 100% of the screen, while others will be contained like a page with a sidebar.

If we amend our template to always render the sidebars, we don't have the option for "if one sidebar do this, if two sidebars do that, if no sidebars do someting else, etc". I think we could perhaps use JS to check and add classes for us, but that might be a little flaky.

Update: it might be worth looking at [this patch](https://git.drupalcode.org/project/dxpr_theme/-/commit/7810b0c78d866e70525729212410a5d6d81a6ce9) to see if it fixes our use case. The authors say it works for 99.9% of their use cases.

---
Thanks to [Big Blue Door](https://www.bigbluedoor.net/) for sponsoring my time to work on this.

**Comment by ctorgalson** [comment]
@markconroy 

> Update: it might be worth looking at this patch

I think you missed the link :)

> We are using the node template to add the container class for nodes, so some nodes or items in nodes can take up 100% of the screen, while others will be contained like a page with a sidebar.

This is what I was trying (badly) to describe on one of the Technical Drop-in calls. The general `:has()` approach doesn't work particularly well with utility classes like `container`. It really _can't_, because we're currently outputting those classes on the basis of the `has_sidebar_*` classes that are the source of the original issue.

For it to work, we'd need something like this:

```css
/* Assumes we *always* output the div.container wrapper in Twig. */

/* Container has *no* sidebar content. */
.lgd-container {
  margin: 0;
  max-width: 100%;
}

/* Container has *any* sidebar content. */
.lgd-container:has(.lgd-row .region) {
  margin: 0 auto;
  max-width: var(--width-container);
}
```

> I think we could perhaps use JS to check and add classes for us

If we _are_ considering js, I'd suggest [the staged approach I mentioned above](https://github.com/localgovdrupal/localgov_base/issues/597#issuecomment-2333482094): `has_sidebar_*` vars for most cases, `:has()` for cases where `has_sidebar_*` doesn't work, and `js` _explicitly_ targeted at browsers that don't support `:has()` for all other cases:

- `has_sidebar_*` in Twig: works almost all the time, but occasionally outputs empty sidebars,
- `:has()`: usually works when `has_sidebar_*` fails, but occasionally isn't supported by the browser,
- `js` in non-`:has()` supporting browsers: usually works when `:has()` fails, but users may occasionally not have js enabled
- **empty sidebars displayed**: but _only_ when all above options have failed

(Note that this would still involve _not_ using some of the utility classes).

**Comment by markconroy** [comment]
Link updated now

https://git.drupalcode.org/project/dxpr_theme/-/commit/7810b0c78d866e70525729212410a5d6d81a6ce9

**Comment by erikerskine** [comment]
@markconroy is `has(.region)` enough? Do we not need to look for things _inside_ the region?

If a block is placed in a region, we get something like this:   (ignoring inner divs and main content etc for now)

```
<div class="page">          <-- put here by page.html.twig
  <div class="region">      <-- put here by region.html.twig
    <placeholder>           <-- put here by region.html.twig via its {{ content }} variable
  </div>
</div>
```

which turns into this, if the block produces content:

```
<div class="page">
  <div class="region">
    <div class="block">...</div>    <-- put here by block.html.twig
  </div>
</div>
```

or this, if the block turns out to be empty:

```
<div class="page">
  <div class="region">
                             <-- the placeholder disappeared
  </div>
</div>
```

If no blocks are placed, or we we know ahead of time that they aren't going to be shown, then `region.html` and it's accompanying `<div class="region">` are omitted altogether:

```
<div class="page">
</div>
```

**Comment by andybroomfield** [comment]
Just to note that GDS have updated guidelines to read Safari 15.6 as minimum (as of September 2024).

https://www.gov.uk/service-manual/technology/designing-for-different-browsers-and-devices

<img width="566" alt="Image" src="https://github.com/user-attachments/assets/5f37586c-4b00-48bc-ba3e-bbfb84b9e2d9" />

**Comment by ctorgalson** [comment]
I don't know that this is foolproof, but this _seems_ to do ok in Twig:

```twig
{# Evaluates to 1 if some other key than `#cache` or `#weight` exists, 0 otherwise. #}
{% set has_some_content = some_content|filter((v, k) => k not in ["#cache", "#weight"])|length ? 1 : 0 %}

{% if has_some_content %}
  <div class="some__content">
    {{ some_content }}
  </div>
{% endif %}
```

~I _think_ it should always work because the Twig has to be re-rendered any time the render array changes, but I could be wrong :)~ (see below).

`:has()` [is also supported in all the listed Gov.UK browsers](https://caniuse.com/?search=%3Ahas), though a Twig/PHP solution that doesn't output the content is slightly preferable (e.g. because, if we use `:has()`, we have to do things like `.sidebar:not(:has(> *)) { display: none; }` and that can be accidentally overridden in child themes...)

Edited:

The `filter()` above doesn't account for the variable being undefined/not iterable (i.e. not a render array). Providing an empty default fixes _half_ that problem, but if a variable contains a string, it's all over.

In a macro, we can check the type in a more sophisticated way and without it being such a crazy one-liner, but macros only **render** (never return), so any result like `1` or `0` is always `Twig\Markup`. But if we don't render _at all_ from the macro, we get an empty string, `""`, which is falsy _enough_ for this purpose:

```twig
{#
  Given a variable:
  
    - if non-empty and iterable and any (presumed) renderable array key is found, outputs 1
    - if non-empty and not iterable, interprets the variable as a string and outputs 1
    - outputs nothing in any other case.

  The output from the macro will be either of:

    - a truthy "Twig\Markup" render array of the number `1`
    - a falsy ""

  Sometimes Twig is annoying.

  Usage:

  {% from 'macros.html.twig' import has_content as has_content %}
  {% set has_something %}{{ has_content(some_render_array) }}{% endset %}

  {% if has_something %}
    <div class="something>
      {{ some_render_array }}
    </div>
  {% endif %}
#}
{%- macro has_content(variable, non_render_keys = ["#cache", "#weight"] ) -%}
  {%- if variable is not empty -%}
    {%- if variable is iterable -%}
      {{ (variable|filter((v, k) => k not in non_render_keys)|length) ? 1 : null }}
    {%- else -%}
      1
    {%- endif -%}
  {%- endif -%}
{%- endmacro -%}
```
This works "ok", but it's not great:

```
0 array:3 [▼
  "field_without_content" => ""
  "string_with_content" => Twig\Markup {#8791 ▶}
  "string_without_content" => ""
]
```
If a theme _shipped_ with it, it'd be reasonably safe. But it's error-prone when edited since losing even one of those whitespace-controlling `-` characters causes the macro to render a non-empty string.

Still, if the overall approach is trustworthy (any opinions?), PHP would do ok (maybe as a filter). But [given how long this issue has existed](https://www.drupal.org/project/drupal/issues/953034), I bet this isn't actually any good as a solution...

**Comment by erikerskine** [comment]
> `{% set has_some_content = some_content|filter((v, k) => k not in ["#cache", "#weight"])|length ? 1 : 0 %}`

This issue was raised to address blocks in sidebar regions, and whether those sidebars are empty. Drupal renders blocks lazily, so we need to consider something like `some_content = ['#lazy_builder' => ...]`. The above code would evaluate to `1` but you might still end up with an empty sidebar.

The template, when run, turns `['#lazy_builder' => ...]` into `<drupal-render-placeholder .../>`. Later on (much later!) `<drupal-render-placeholder />` is swapped out for something else.

Is `<drupal-render-placeholder>` empty? Maybe, maybe not. We can only know for sure by evaluating it, but to do that goes directly against what lazy rendering is all about.

This isn't a technical problem. Yes, it's been around a long time, but I don't think it will ever move forward without making a conscious choice between:

1. Accept that "has content" really means "might have content". You get the performance benefits of lazy rendering.
2. Do not accept that. Guarantee empty vs non-empty, but with a performance/caching cost.

**Comment by ctorgalson** [comment]
Thanks @erikerskine, that's nicely put.

I'm for using `:has()` then, though presumably in a forthcoming `localgov_base` theme where we can live with potentially breaking changes.

It's pretty easy. We define a utility class like this:

```css
.hide-when-empty:not(:has(> *)) {
  display: none;
}
```

...then layout grids can do this:

```css
.grid:has(> .hide-when-empty):not(:has(> .hide-when-empty > *)) {
  --grid-template-columns: 1;
}
```

This is quite similar, in terms of usage in `.twig` to what we do now.

Demo: https://codepen.io/bedlamhotel/pen/wBMamdv?editors=1100

@markconroy @tonypaulbarker

**Comment by markconroy** [comment]
I think I'm coming around to the `:has()` approach as well, at least until Drupal figures out a way to solve this for us.
